### ISSUE TITLE
Node killed by OOM 

### ISSUE THREAD

**Issue by leomkkwan** [cockroachdb/cockroach#54360]
It’s a 10 nodes cluster, running on gcp with this config n2-standard-8 (8 vCPUs, 32 GB memory)
It is running on Ubuntu 18.04 without Kubernetes.
The problem is exists on both V20.2.0-ALPHA.3 and V20.1.5..

All nodes start with this command 

ulimit -n 65536 && ./cockroach start --cache=25%  --max-sql-memory=25%  --background --locality=cloud=gcp,region=us-east1,datacenter=us-east1-c --store=path=/mnt/d1,attrs=ssd,size=90% --log-dir=log --certs-dir=certs --max-disk-temp-storage=100GB --locality-advertise-addr=cloud=gcp@172.28.0.29,region=us-east1@172.28.0.29,datacenter=us-east1-c@172.28.0.29 --join=172.28.0.29,172.28.0.30,172.28.0.31,172.28.0.32,172.28.0.33,172.28.0.34,172.28.0.35 --advertise-addr=172.28.0.29

Here is the schema and query

CREATE TABLE t_baseid (
             bid INT8 NOT NULL,
             emailcrc INT8 NOT NULL,
             emailmd5 STRING(32) NOT NULL,
             emailsha1 STRING(40) NOT NULL,
             email STRING NOT NULL,
             CONSTRAINT "primary" PRIMARY KEY (bid ASC),
             INDEX idx_emailsha1 (emailsha1 ASC),
             INDEX idx_emailcrc (emailcrc ASC),
             INDEX idx_emailmd5 (emailmd5 ASC),
             FAMILY "primary" (bid, emailcrc, emailmd5, emailsha1, email)

root@:26257/filu> explain select email,count(0) from t_baseid group by email having count(0) > 1;
         tree         |    field    |   description
----------------------+-------------+-------------------
                      | distributed | true
                      | vectorized  | false
  filter              |             |
   │                  | filter      | count > 1
   └── group          |             |
        │             | aggregate 0 | email
        │             | aggregate 1 | count(column6)
        │             | group by    | email
        └── render    |             |
             └── scan |             |
                      | table       | t_baseid@primary
                      | spans       | FULL SCAN
(12 rows)

Time: 157.441488ms

The oom issue happened on a v20.1.5 cluster on the 4th try, but it could also happen on the first try. 

root@:26257/filu> select email,count(0) from t_baseid group by email having count(0) > 1;
ERROR: root: memory budget exceeded: 1073741824 bytes requested, 7883292672 currently allocated, 8419290112 bytes in budget
SQLSTATE: 53200
root@:26257/filu> select email,count(0) from t_baseid group by email having count(0) > 1;
ERROR: root: memory budget exceeded: 1073741824 bytes requested, 7930335232 currently allocated, 8419290112 bytes in budget
SQLSTATE: 53200
root@:26257/filu> select email,count(0) from t_baseid group by email having count(0) > 1;
ERROR: root: memory budget exceeded: 1073741824 bytes requested, 8021225472 currently allocated, 8419290112 bytes in budget
SQLSTATE: 53200
root@:26257/filu> select email,count(0) from t_baseid group by email having count(0) > 1;
ERROR: driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the transaction status: dial tcp 127.0.0.1:26257: connect: connection refused
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the database name: dial tcp 127.0.0.1:26257: connect: connection refused
root@:26257/? ?>

**Comment by awoods187** [comment]
Thank you for the bug report--we will take a look!

**Comment by yuzefovich** [comment]
I've been looking into trying to reproduce this issue for a bit, and I haven't been able to, but I do see something suspicious.
![Screen Shot 2020-09-17 at 2 57 21 PM](https://user-images.githubusercontent.com/5135385/93532549-19fee580-f8f6-11ea-9c77-022e9b173db5.png)
In particular, I think `rowenc.DecodeUntaggedDatum` is suspicious - what happens here is that we get `EncDatum` with `encoded` field set from the table reader, then we need to decode it when performing `Fingerprint` (computing the hash of grouping columns), so we allocate a bunch of strings, those allocations aren't accounted anywhere. But what confuses me is that why those rows are not garbage collected - we're losing references to them once we perform the aggregation. I tried this diff
```
diff --git a/pkg/sql/rowexec/aggregator.go b/pkg/sql/rowexec/aggregator.go
index 3bf7d043de..f9c51d091b 100644
--- a/pkg/sql/rowexec/aggregator.go
+++ b/pkg/sql/rowexec/aggregator.go
@@ -803,6 +803,7 @@ func (ag *aggregatorBase) accumulateRowIntoBucket(
                                continue
                        }
                        otherArgs[j-1] = row[c].Datum
+                       row[c].UnsetDatum()
                }
```
but still observed the same behavior. From what I can see we're doing a good job estimating the memory usage of the aggregation (apart from this `DecodeUntaggedDatum`).

My current thinking is that we're not accounting for several things (`DecodeUntaggedDatum` at SQL level and possibly for reads at KV level) which combined with slow GC (or a delay in GC) could lead to a crash. The mitigation could be lowering `--max-sql-memory` parameter from recommended to 25% so that we would hit the memory error sooner in this case, but this is a poor solution. I'll continue looking into this.

**Comment by yuzefovich** [comment]
Following @jordanlewis suggestion to use `runtime.SetFinalizer` I was able to verify that those unaccounted-for allocations in `rowenc.DecodeUntaggedDatum` take a while to be garbage-collected. Here is what I'm seeing:
- I'm executing a query with hash aggregation and the memory usage constantly increases (as expected):
```
I200923 21:25:40.791909 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 143 MiB RSS, 368 goroutines, 50 MiB/67 MiB/82 MiB GO alloc/idle/total, 54 MiB/67 MiB CGO alloc/total, 114.6 CGO/sec, 5.2/2.2 %(u/s)time, 0.0 %gc (3x), 1.3 MiB/1.3 MiB (r/w)net
I200923 21:25:50.868206 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 2.8 GiB RSS, 368 goroutines, 1.9 GiB/41 MiB/2.2 GiB GO alloc/idle/total, 617 MiB/705 MiB CGO alloc/total, 6025.0 CGO/sec, 176.6/9.0 %(u/s)time, 0.0 %gc (7x), 1.3 MiB/1.3 MiB (r/w)net
I200923 21:26:00.839161 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 3.5 GiB RSS, 370 goroutines, 2.3 GiB/34 MiB/2.7 GiB GO alloc/idle/total, 842 MiB/958 MiB CGO alloc/total, 2298.3 CGO/sec, 161.2/6.8 %(u/s)time, 0.0 %gc (1x), 1.3 MiB/1.4 MiB (r/w)net
```
- however, once the query is done, the cluster becomes idle and the memory usage remains elevated (I believe GC hasn't been run, so we see a big number in `Go alloc`):
```
I200923 21:26:20.860200 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 3.5 GiB RSS, 369 goroutines, 2.4 GiB/5.6 MiB/2.7 GiB GO alloc/idle/total, 842 MiB/959 MiB CGO alloc/total, 1.8 CGO/sec, 2.4/1.9 %(u/s)time, 0.0 %gc (0x), 1.3 MiB/1.3 MiB (r/w)net
I200923 21:26:30.838078 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 3.5 GiB RSS, 366 goroutines, 2.4 GiB/54 MiB/2.7 GiB GO alloc/idle/total, 842 MiB/959 MiB CGO alloc/total, 0.6 CGO/sec, 3.1/2.8 %(u/s)time, 0.0 %gc (0x), 1.4 MiB/1.4 MiB (r/w)net
I200923 21:26:40.846768 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 3.5 GiB RSS, 368 goroutines, 2.4 GiB/41 MiB/2.8 GiB GO alloc/idle/total, 842 MiB/959 MiB CGO alloc/total, 0.6 CGO/sec, 2.3/1.9 %(u/s)time, 0.0 %gc (0x), 1.3 MiB/1.3 MiB (r/w)net
```
- only once I try to execute the same query again do I see the finalizers called (meaning the datums allocated by `DecodeUntaggedDatum` are attempted to be GC'ed - and they would have been garbage-collected if it were not for the finalizers I added):
```
I200923 21:27:10.853992 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 3.5 GiB RSS, 369 goroutines, 2.4 GiB/2.0 MiB/2.8 GiB GO alloc/idle/total, 843 MiB/960 MiB CGO alloc/total, 0.6 CGO/sec, 2.2/2.1 %(u/s)time, 0.0 %gc (0x), 1.2 MiB/1.3 MiB (r/w)net
I200923 21:27:20.071348 18 sql/rowenc/column_type_encoding.go:553 ⋮ finalizer 2020-09-23 14:27:20.071334 -0700 PDT m=+119.912943330
I200923 21:27:20.084521 18 sql/rowenc/column_type_encoding.go:553 ⋮ finalizer 2020-09-23 14:27:20.08452 -0700 PDT m=+119.926129075
...
```
If I leave the cluster alone (meaning no load) for several minutes, eventually GC will reclaim the memory and we'll be in a happy state:
```
I200923 21:32:40.951929 455 server/status/runtime.go:522 ⋮ [n1] runtime stats: 5.1 GiB RSS, 368 goroutines, 37 MiB/3.8 GiB/3.6 GiB GO alloc/idle/total, 878 MiB/998 MiB CGO alloc/total, 0.6 CGO/sec, 5.3/7.4 %(u/s)time, 0.0 %gc (0x), 1.2 MiB/1.3 MiB (r/w)net
```
----
I believe that my initial hypothesis (mentioned at the very end of the previous comment) for why @leomkkwan is able to easily crash the cluster is correct - it is a combination of a few things:
- 1) we do not account for the datums that might get allocated in `Fingerprint` (which is a hashing function used to perform the grouping in case of hash aggregation). During my investigation, that memory would take up about 25% of all memory used for aggregation, so it's a pretty significant.
- 2) I believe not accounting in step 1) would have been acceptable if Go GC was a bit more eager to collect the memory under no-longer referenced objects. Those allocations in `Fingerprint` are only temporary (they are only needed for a split second), and then we're losing references to them pretty much right away, but GC takes its time reclaiming the memory.
- 3) what could makes things even worse is that we might have automatic statistics collection running at the same time (assuming that a freshly created cluster with recently inserted data). The stats collection also uses `Fingerprint` method to sample the rows.

So now the question is how do we improve in such scenario. Note that currently because of the lack of disk spilling in hash aggregation (#42485), the desired behavior is hitting a memory out of budget error instead of a crash. I see the following options (we might decide to use a combination of these):
1) we should account for the datums allocated during `Fingerprint` evaluation. Although we're losing the references to them, the memory is likely to not be reclaimed for some time, so we would benefit stability-wise if we add the accounting (ditto for stats collection)
2) we need to allocate those datums in the first place only because we want to encode them differently (they are coming in with value encoding, and we request key ASC encoding, so we decode from value to `tree.Datum` and then encode it using key ASC). We might relax those requirements and be more flexible removing the decoding-encoding step altogether.
3) we might consider explicitly calling `runtime.GC()` once a big hash aggregation is over to nudge the GC. I'd expect this to be a no-go though.
----
Note that what I described above is mostly relevant to the row-by-row engine, and the vectorized engine should be better equipped in such situations (it has much better memory accounting story as well as used a different hashing technique), so @leomkkwan I'd be interested if you could try the queries with `SET vectorize=on`. Note that such setting will be the default on 20.2 release, but the stats might be missing, in which case we would fallback to using the old row-by-row engine (to go around the stats issue you could do `SET vectorize_row_count_threshold=0`). The vectorized engine also isn't able to spill the hash aggregation to disk, so I'd expect that you'd be hitting memory budget errors. That said, we plan on adding that ability in 21.1 which should come out next Spring.

**Comment by yuzefovich** [comment]
Duh - I missed something very obvious: the query itself (apart from grouping on `email` column) returns that `email` column as well, so we do keep the reference to the decoded datums until the query execution finishes. I'll update the comment above accordingly, but it still remains valid in many regards.

**Comment by leomkkwan** [comment]
Thank you for look into this. 7 out of 10 nodes get killed when I turn on vectorize and run the same query.

```
root@:26257/filu>
SET vectorize=on;
SET

Time: 185.069205ms

root@:26257/filu> show vectorize;
  vectorize
-------------
  on
(1 row)

Time: 728.556µs

root@:26257/filu> explain select email,count(0) from t_baseid group by email having count(0) > 1;
         tree         |    field    |   description
----------------------+-------------+-------------------
                      | distributed | true
                      | vectorized  | true
  filter              |             |
   │                  | filter      | count > 1
   └── group          |             |
        │             | aggregate 0 | email
        │             | aggregate 1 | count(column6)
        │             | group by    | email
        └── render    |             |
             └── scan |             |
                      | table       | t_baseid@primary
                      | spans       | FULL SCAN
(12 rows)

Time: 862.891027ms

root@:26257/filu> select email,count(0) from t_baseid group by email having count(0) > 1;
ERROR: driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the transaction status: dial tcp 127.0.0.1:26257: connect: connection refused
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the database name: dial tcp 127.0.0.1:26257: connect: connection refused
root@:26257/? ?>


```

**Comment by yuzefovich** [comment]
Hm, I'm a bit puzzled to be honest - I still haven't encountered a crash when investigating locally (maybe I'm just working with not big enough dataset). How many rows are in `t_baseid` table? How many different `email`s are in there? Do you know what is the average length of a value in `email` column?

Also, if you're comfortable sharing it, it'd be awesome if you could collect a [debug.zip](https://www.cockroachlabs.com/docs/v20.1/cockroach-debug-zip) and send it over to yahor@cockroachlabs.com (if full debug.zip is a bit too much, heap profiles would be great too - they are in `<store>/logs/heap_profiler`).

**Comment by yuzefovich** [comment]
Hm, I tried implementing suggestion 2) (being more flexible with encoding we use) in #54722, but the approach turns out to not work easily because the value encoding contains somewhat non-deterministic information in the value tag component (the contents might different based on the column family specification and whatnot), so I'll abandon that prototype. I'll still add the memory accounting in a separate PR.

**Comment by yuzefovich** [comment]
I'll leave this issue aside for now waiting for @leomkkwan to provide some more information as requested [here](https://github.com/cockroachdb/cockroach/issues/54360#issuecomment-698437561).

**Comment by leomkkwan** [comment]
Hi @yuzefovich , 
Another test was done on v20.1.6. Three nodes ,1, 5 and 11 were killed in the test.

```

root@:26257/filu> SET vectorize=on;
SET

Time: 343.749µs

root@:26257/filu> select now(); select email,count(0) from t_baseid group by email having count(0) > 1; select now();
ERROR: driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the transaction status: dial tcp 127.0.0.1:26257: connect: connection refused
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the database name: dial tcp 127.0.0.1:26257: connect: connection refused
root@:26257/? ?> ^D
ERROR: driver: bad connection
Failed running "sql"
root@cockroach-01:/cockroach# date
Fri Oct  2 13:21:59 EDT 2020

```

Debug file is sent to you through email. Thanks.

**Comment by yuzefovich** [comment]
Thanks for sending that information over.

Looking at the node 1 I see the following as the last runtime stats (I'm assuming the node dies shortly, within a couple of seconds):
```
I201002 17:21:24.702234 317 server/status/runtime.go:498  [n1] runtime stats: 30 GiB RSS, 600 goroutines, 20 GiB/1.2 GiB/23 GiB GO alloc/idle/total, 7.9 Gi    B/9.5 GiB CGO alloc/total, 179001.3 CGO/sec, 263.0/57.1 %(u/s)time, 0.0 %gc (0x), 28 MiB/281 MiB (r/w)net
```
And the heap profile taken at the same time shows this:
![Screen Shot 2020-10-02 at 11 12 46 AM](https://user-images.githubusercontent.com/5135385/94955937-34ad8e80-04a0-11eb-979b-6e2946cae7c1.png)

So out of 20GB reported in use by Go we can attribute almost 10GB to the hash aggregation. Note that `max-sql-memory` is set to 25% which comes out to 7.8GB which means that even if we ignore the question "where do the second 10GB of RAM go?", we are still expected to run into a memory budget error. This observation shows that we have not accounted for everything in the vectorized hash aggregator.

One obvious thing is that on 20.1 we use Golang's `map` for some bookkeeping and it is not accounted for. I'm having some issues parsing the profile (I'm on Mac, and the profile was taken on Linux), but that unaccounted memory contributes to 1.6GB that are allocated within `onlineAgg` itself and other functions it calls, and it most it should be on the order of 1GB. This part has been refactored in 20.2 (that `map` was entirely removed).

That still leaves us at 9GB or so of what-should-be-accounted-for memory. We do seem to have the accounting in all other big memory usages, but we're probably under-accounting for some reason. As a side comment, based on 763MB coming from `newHashAggFuncs` we can make the following computation: `# of groups = 763MB / 2KB * 64` which is approximately 25 million (`hashAggFuncsAllocSize = 64`).

**Comment by yuzefovich** [comment]
I see another issue that I don't think we have addressed on the master branch - we probably don't account for 722MB in `anyNotNullBytesAgg.Compute`. What happens there is that we store `curAgg` value of size of 32 bytes for each group (so that we can display `email` for the query) on `anyNotNullBytesAgg` struct, and we don't account for that memory. We store that data temporarily until the aggregate struct finds another group; however, in case of the hash aggregation it will never see another group until the whole aggregation has been complete, so we end up holding onto all of that data throughout the whole execution and it goes unaccounted for. `Bytes` and `String` types are the only susceptible to this issue on 20.1 branch because they are of variable size. On 20.2 we'll additionally need to be careful with `tree.Datum` backed types.

**Comment by yuzefovich** [comment]
Another thing we don't account for is the overhead of a slice (32 byte allocations) in `makeAggregateFuncs` which show another 755MB. The assumption was that it was negligible, but as we can see that's not always the case.

An interesting observation from my local debugging is that I think I have a case when Go profiler is quite inaccurate. Consider this snippet:
```
         .     1.73GB    454:			op.keyMappingAllocator.PerformOperation(op.keyMapping.ColVecs(), func() {
         .          .    455:				for keyIdx, colIdx := range op.groupCols {
         .          .    456:					// TODO(azhng): Try to preallocate enough memory so instead of
         .          .    457:					// .Append() we can use execgen.SET to improve the
         .          .    458:					// performance.
         .          .    459:					op.keyMapping.ColVec(keyIdx).Append(coldata.SliceArgs{
         .          .    460:						Src:         op.scratch.ColVec(int(colIdx)),
         .          .    461:						ColType:     op.inputTypes[colIdx],
         .          .    462:						DestIdx:     aggFunc.keyIdx,
         .          .    463:						SrcStartIdx: remaining[0],
         .          .    464:						SrcEndIdx:   remaining[0] + 1,
         .          .    465:					})
         .          .    466:				}
         .          .    467:				op.keyMapping.SetLength(keyIdx + 1)
         .          .    468:			})
         .          .    469:
         .       14MB    470:			aggFunc.fns, _ = makeAggregateFuncs(op.aggFuncsAllocator, op.aggTypes, op.aggFuncs)
         .          .    471:			op.aggFuncMap[hashCode] = append(op.aggFuncMap[hashCode], aggFunc)
         .          .    472:
         .          .    473:			// Select rest of the tuples that matches the current key. We don't need
         .          .    474:			// to check if there is any match since 'remaining[0]' will always be
         .          .    475:			// matched.
         .          .    476:			// TODO(azhng): Refactor match so that we can skip checking remaining[0].
         .          .    477:			_, remaining = aggFunc.match(
         .          .    478:				remaining, op.scratch, op.groupCols, op.groupTypes, op.keyMapping,
         .          .    479:				op.scratch.group[:len(remaining)],
         .          .    480:			)
         .          .    481:
         .          .    482:			// Hack required to get aggregation function working. See '.scratch.group'
         .          .    483:			// field comment in hashAggregator for more details.
         .          .    484:			op.scratch.group[groupStartIdx] = true
         .          .    485:			aggFunc.init(op.scratch.group, op.output)
         .   806.83MB    486:			aggFunc.compute(op.scratch, op.aggCols)
```
The two big allocations are actually storing exactly the same amount of data albeit with a different layout:
- 1.73GB are stored contiguously as a single huge `[]byte`
- 806MB are stored in many separate structs as separate small `[]byte` (of about length 10000)

The two numbers should grow roughly the same. Of course, there might be a difference in that for the large slice Go preallocate a bigger underlying memory array, so the large slice `cap` is much greater than `len` whereas for many smaller slices the overhead is smaller, but I'd doubt it should have 2x difference.

----

Anyway, @leomkkwan I think I finally have an explanation to why we're seeing the crashes with the vectorized engine. There are a couple of factors to blame:
- we're scanning a lot of data below SQL layer (which is not registered with our memory accounting system), and although that data is being used only temporarily, Go GC is quite slow to reclaim the memory. That is especially important given that the memory usage under the vectorized hash aggregation grows very rapidly.
- we are not accounting for all the memory used by the hash aggregation (in three places - usage `map`, in `anyNotNullAgg.curAgg`, and the overhead of a slice in `makeAggregateFuncs`). What makes things worse is that internally `map` might be reallocated a few times, so its memory footprint might actually be larger than what shown at the profile.

The good news:
- the hash aggregation has been refactored in 20.2, the usage of `map` has been removed
- I'll fix the missing accounting (hopefully we'll get it into 20.2.0)
- we will address the current limitation of the hash aggregation of not being able to spill to disk in 21.1
- we added some memory accounting below SQL layer for scan in 20.2

The bad news:
- I think even with perfect memory accounting, it'll still be possible to crash the cluster when such queries are executed in quick succession (i.e. execute -> get memory budget error -> execute -> get memory budget error -> ... -> execute -> a crash) due to the slowness of the garbage collector. This will be the case until we have the disk spilling support in the hash aggregation.
